### PR TITLE
TOOLS: StereoImageSubscriber, ImageMux, misc

### PR DESCRIPTION
**Pull request by kev-the-dev**
* add ImageMux, a class for creating a grid of images (in python), for doing things like combining many debug images into one
* add StereoImageSubscriber, the stereo analog of the ImageSubscriber in mil_tools. Abstracts image syncing and ROS -> numpy conversion so devs can simply provide a callback and start using stereo (in python)
* some misc other stuff (see diff or commit messages)

### CODE REVIEW

**Comment by kev-the-dev** [comment]
The power is now yours. Try ImageMux today!
imagemux

File: perception/mil_vision/mil_vision_tools/image_mux.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+class ImageMux(object):
+    '''
+    Utility to create a customizable square grid of images, with labels and borders,
+    which will generate a single image from the grid at any time. Useful for
+    combining several debug images into one.
+
+    See bottom of this file for a usage example.
+
+    TODO: implement keep_ratio as described below
```
Comment (kingkevlar) [inline perception/mil_vision/mil_vision_tools/image_mux.py]: It's a little inconvenient to not have the option to retain the ratio when I tested this.

File: perception/mil_vision/mil_vision_tools/image_mux.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+if __name__ == '__main__':
+    '''
+    ImageMux is intended to be used as a class, not an executable.
+    The following is an example of how to use it in a python program.
+    Creates a 2x2 grid of Racoon images with labels, using some custom parameters.
+
+    To run this yourself, download some images, put them in $HOME/Pictures/[1.jpg, 2.jpg, 3.jpg, 4.jpg]
+    '''
+    import os
+    labels = ['Chubby Racoon', 'Kiddo Racoons', None, 'Racoons in Da Street']  # Create strings for labels
```
Comment (kingkevlar) [inline perception/mil_vision/mil_vision_tools/image_mux.py]: RacoonMux?

File: utils/mil_tools/mil_ros_tools/image_helpers.py [hunk 3]
```diff
@@ -121,3 +122,103 @@ def convert(self, data):
...
+        @param right_image_topic ROS topic to subscribe to for the right camera ex: /camera/front/right/image_rect_color
+        @param callback Function with signature foo(left_img, right_img) to call when a synchronized pair is ready.
+               If left as None, the latest synced images are stored as self.image_left and self.image_right
+        @param slop Maximum time in seconds between left and right images to be considered synced.
+               If left as None, will only consider synced if left and right images have exact same header time.
+        @param encoding String to pass to CvBridge to encode ROS image message to numpy array
+        @param queue_size Integer, the number of images to store in a buffer for each camera to find synced images
+        '''
+        if callback is None:  # Set default callback to just set image_left and image_right
+            def callback(image_left, image_right):
```
Comment (kingkevlar) [inline utils/mil_tools/mil_ros_tools/image_helpers.py]: In keeping sync with the other ImageSubscriber I think'last_image_left' and 'last_image_right' would be more natural.

File: utils/mil_tools/mil_ros_tools/image_helpers.py [hunk 4]
```diff
@@ -121,3 +122,103 @@ def convert(self, data):
...
+                return self.camera_info_left, self.camera_info_right
+            rospy.sleep(0.05)
+        if self.camera_info_left is not None and self.camera_info_right is not None:
+            if unregister:
+                self._info_sub_left.unregister()
+                self._info_sub_right.unregister()
+            return self.camera_info_left, self.camera_info_right
+        raise Exception("Camera info not found.")
+
+    def _image_callback(self, left, right):
```
Comment (kingkevlar) [inline utils/mil_tools/mil_ros_tools/image_helpers.py]: left_img would be more immediately parsed (or add to docstring like all of your other cute docstrings.)

Comment (kingkevlar) [inline utils/mil_tools/mil_ros_tools/image_helpers.py]: left_img and right_img would be more immediately parseable. Else, I'd add it to the docstring.

File: perception/mil_vision/mil_vision_tools/image_mux.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        self.border_color = border_color
+        self.border_thickness = border_thickness
+        self.text_color = text_color
+        self.text_font = text_font
+        self.text_scale = text_scale
+        self.text_thickness = text_thickness
+
+        # If labels not specified, fill a list with None's
+        if labels is None:
+            self.labels = [None for _ in xrange(self.shape[0] * self.shape[1])]
```
Comment (DSsoto) [inline perception/mil_vision/mil_vision_tools/image_mux.py:44]: `[None, ] * self.shape[0] * self.shape[1]`?

File: perception/mil_vision/mil_vision_tools/image_mux.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        '''
+        assert isinstance(img, np.ndarray), 'img must be numpy array'
+        # If image is grayscale, convert to 3 channel
+        if len(img.shape) == 2 or img.shape[2] == 1:
+            img = cv2.cvtColor(img, cv2.COLOR_GRAY2BGR)
+
+        if isinstance(key, int):  # Accept a single index, ex 5 -> (2, 1)
+            key = self._index_to_tuple(key)
+        assert isinstance(key, tuple), 'must be tuple'
+        assert len(key) == 2, 'index best be 2D'
```
Comment (DSsoto) [inline perception/mil_vision/mil_vision_tools/image_mux.py:107]: Haha 'best be' sounds pretty sassy.

**Review by DSsoto — APPROVED** [review]

**Review by kingkevlar — APPROVED** [review]

File: perception/mil_vision/mil_vision_tools/image_mux.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        '''
+        Sets the content of one pane in the image grid.
+        @param The index of the pane to set to the data of img
+               If an integer -> sets pane at index, counting left to right, then top to bottom
+               If a tuple (row, col) -> set pane at specified (index 0) row and column
+        @param img numpy array with shape (m, n, 3) or (m, n, 1) representing the image to insert
+               in the pane specified in key. If a one channel image, first convert grayscale to BGR.
+               If keep_ratio was True in contructor, will add black bars as necessary to fill pane.
+               Otherwise, use standard cv2.resize to fit img into pane.
+        @throw AssertionError if key is wrong type of out of bounds
```
Comment (DSsoto) [inline perception/mil_vision/mil_vision_tools/image_mux.py]: It's raise, not throw

**Pull request merged**